core/src/main/java/org/apache/calcite/plan/volcano/VolcanoRuleMatch.java

```diff
@@ -144,7 +144,10 @@ private String computeDigest() {
       if (i > 0) {
         buf.append(", ");
       }
-      buf.append(rels[i].toString());
+      // Same with {@link rel#getDescription()}, we don't call it explicitly
+      // in order to avoid string copy when generating description.
+      buf.append("rel#").append(rels[i].getId())
+          .append(':').append(rels[i].getDigest());
```

Comment:
Can we abstract this description as a tool method and reuse that ? Something like `RelOptUtil#generatesDescription(StringBuilder builder, int relNodeID, String digest)`

Reply:
Sounds good. I have addressed your comments.